api/query/cache/service/main.go

```diff
@@ -85,6 +87,55 @@ func searchHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	//
+	// Start: Shim to ignore irrelevant tests.
+	//
+
+	// Interpret execution plan as a concrete query that can be manipulated.
+	q, ok := plan.(query.ConcreteQuery)
+	if !ok {
+		http.Error(w, "Failed bind test runs to abstract query", http.StatusInternalServerError)
+		return
+	}
+
+	// Create base query of the form
+	// OR(!run1-status:UNKNOWN, ..., !runN-status:UNKNOWN).
+	baseQuery := query.Or{
+		Args: make([]query.ConcreteQuery, len(rq.RunIDs)),
+	}
+	for i, runID := range rq.RunIDs {
+		baseQuery.Args[i] = query.Not{
+			Arg: query.RunTestStatusConstraint{
+				Run:    runID,
+				Status: shared.TestStatusUnknown,
+			},
+		}
+	}
+
+	// Add baseQuery to existing AND in q=AND(...), or create AND(baseQuery, q).
+	if andQ, ok := q.(query.And); ok {
+		andQ.Args = append([]query.ConcreteQuery{baseQuery}, andQ.Args...)
+		q = andQ
```

Comment:
Nit: Do you expect this optimization to matter much? If not, I don't see the harm in a tree that's 1 layer deeper.

Reply:
It could matter because we will eventually want to optimize the order in which predicate fragments run. This optimization is much easier to do heuristically if we keep filters as "flat" as possible.

The other reason is consistency: Other `AND`/`OR` managing code contains these checks to keep queries "flat".